service/history/workflow/mutable_state_impl.go

```diff
@@ -863,6 +863,20 @@ func (e *MutableStateImpl) GetCompletionEvent(
 	return event, nil
 }
 
+// GetWorkflowCloseTime returns workflow closed time, returns nil for open workflow
+func (e *MutableStateImpl) GetWorkflowCloseTime(ctx context.Context) (*time.Time, error) {
+	if e.executionState.GetState() == enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED && e.executionInfo.CloseTime == nil {
+		// This is for backward compatible. (Added by 05/21/2022)
```

Comment:
Maybe add some details here like workflow closed by 1.16.x server does not have close time in mutable state? and a TODO for removing the logic?